### PR TITLE
Timeout if RN app doesn't boot

### PR DESCRIPTION
**Pull request by JonathanAndrews**
### 🚨 Before Merging
Update the `--boot-timeout` flag description with the required version numbers for Cavy and Cavy Cli. --> https://github.com/pixielabs/cavy-cli/pull/29#discussion_r426549914

### Includes
- add --boot-timeout flag to the cli.
- if the React Native application does not boot within stated time, shut down cavy cli.
- receive communications from Cavy via a web socket.

### Testing
To test you need use this [branch](https://github.com/pixielabs/cavy/tree/timeout-if-app-doesnt-boot) of Cavy.
See this [PR](https://github.com/pixielabs/cavy/pull/177).

### CODE REVIEW

File: cavy.js [hunk 1]
```diff
@@ -55,6 +56,10 @@ program
     'Swap the index files and start the report server without first building the app'
   )
   .option('-d, --dev', 'Keep report server alive until manually killed')
+  .option(
+    '-bt, --boot-timeout <seconds>',
+    'Set how long the CLI should wait for the RN app to boot (defaults to 20 seconds)'
```
Comment (AbigailMcP) [inline cavy.js]: How did you pick the default? Is this default enough time for our sample app to build in CircleCI?

File: cavy.js [hunk 2]
```diff
@@ -55,6 +56,10 @@ program
     'Swap the index files and start the report server without first building the app'
   )
   .option('-d, --dev', 'Keep report server alive until manually killed')
+  .option(
+    '-bt, --boot-timeout <seconds>',
```
Comment (AbigailMcP) [inline cavy.js]: The `--bt` option doesn't seem work - it just uses default 20 seconds. I think it has to be a single letter to work?

If you feel like a shorthand option doesn't feel right, we don't have to have one.

File: server.js [hunk 3]
```diff
@@ -1,9 +1,27 @@
...
+const WebSocket = require('ws');
 const chalk = require('chalk');
 const constructXML = require('./src/junitFormatter');
 
-server.use(express.json());
+// Initialize a server
+const server = http.createServer();
+
+// Setup local variables for server
+server.locals = { appBooted: false};
```
Comment (AbigailMcP) [inline server.js]: This works and we've kept the same `locals` from when we were in express land. I don't know whether there's a 'proper/expected way' to do this for our websocket server - @jalada all good?

File: server.js [hunk 4]
```diff
@@ -1,9 +1,27 @@
...
+
+// Initialize a WebSocket Server instance
+const wss = new WebSocket.Server({server});
+
+// Setup the wanted behaviour for specific socket events
```
Comment (AbigailMcP) [inline server.js]: Can you make this comment a little less generic? i.e. what's the behaviour and what are the events?

File: server.js [hunk 5]
```diff
@@ -1,9 +1,27 @@
...
+wss.on('connection', socket => {
+  socket.on('message', message => {
+    const resultsJSON = JSON.parse(message);
```
Comment (AbigailMcP) [inline server.js]: Just be a style thing, but I'd call this variable `resultsJson` - more consistent with variables we have like `outputAsXml`

File: server.js [hunk 6]
```diff
@@ -14,11 +32,11 @@ function countString(count, str) {
   return `${count} ${string}`;
 }
 
-// Public: POST route which accepts json report object, console logs the results
+// Internal: Accepts a json report object, console logs the results
 // and quits the process with either exit code 1 or 0 depending on whether any
 // tests failed.
-server.post('/report', (req, res) => {
-  const { results, fullResults, errorCount, duration } = req.body;
+function processReport(resultsJSON) {
```
Comment (AbigailMcP) [inline server.js]: Same var name thing as above.

File: src/runTests.js [hunk 7]
```diff
@@ -50,14 +50,21 @@ function getAdbPath() {
...
   server.locals.dev = dev;
   server.locals.outputAsXml = outputAsXml;
   server.listen(8082, () => {
     if (command == 'run-android') {
       runAdbReverse();
     }
     console.log(`cavy: Listening on port 8082 for test results...`);
+    setTimeout(() => {
+      if (!server.locals.appBooted) {
+        console.log("No response from Cavy.");
```
Comment (AbigailMcP) [inline src/runTests.js]: It would be helpful to add the bootTimeout to this message.

File: src/runTests.js [hunk 8]
```diff
@@ -50,14 +50,21 @@ function getAdbPath() {
...
+        console.log("Terminating processes.");
```
Comment (AbigailMcP) [inline src/runTests.js]: I know this is super annoying, but can we change these (and above) to single quotes like the rest of the file?

File: src/runTests.js [hunk 9]
```diff
@@ -108,8 +116,11 @@ function runTests(command, file, skipbuild, dev, outputAsXml, args) {
     process.exit(1);
   });
 
+  // Convert bootTimeout to milliseconds, default to 20 seconds
+  const timeout = (bootTimeout * 1000) || 20000
```
Comment (AbigailMcP) [inline src/runTests.js]: semicolon and also can we move this magic number 20000 to somewhere?

**Review by AbigailMcP — CHANGES_REQUESTED** [review]
Comments inline!

Also, the behaviour seems strange if the user runs `cavy run-ios -s`? This assumes that the app is already built and running, so in this case the timeout is effectively waiting for someone to refresh the app?

File: cavy.js (hunk 1 above)
Comment (AbigailMcP) [inline cavy.js]: To for publishing: we need to add a note about which Cavy version to use here (at the moment not sure what this will be)

Comment (JonathanAndrews) [reply cavy.js]: @jalada What do you think is a reasonable default for the boot timeout?
At the moment, the default is 20 seconds which is only reasonable for the tiny sample app.

File: cavy.js (hunk 2 above)
Comment (jalada) [reply cavy.js]: Why not `-t`? 

Make sure to check none of these clash with the react-native CLI. I dunno if we checked the others though...

File: cavy.js (hunk 1 above)
Comment (jalada) [reply cavy.js]: [CircleCI has a 10 min no-output timeout](https://support.circleci.com/hc/en-us/articles/360007188574-Build-has-hit-timeout-limit). So somewhere between 20 seconds and 10 minutes. How about 2 minutes?

File: server.js (hunk 3 above)
Comment (jalada) [reply server.js]: You could literally just have a global variable. `var appBooted = false;`

File: src/runTests.js [hunk 10]
```diff
@@ -5,8 +5,16 @@ const server = require('../server');
 const { existsSync } = require('fs');
 const { spawn, execFileSync, execSync } = require('child_process');
 
+// This is the default value for boot timeout
+const TWO_MINUTES = 2;
```
Comment (AbigailMcP) [inline src/runTests.js]: but if we change the integer, we have to change the variable name too! in two places! 

```
// Default boot timeout in minutes
const BOOT_TIMEOUT = 2
```

**Review by AbigailMcP — COMMENTED** [review]
Looking good 👍 

I think that now we're ignoring the boot timeout when used with the skip build option, we should console log a message to the user if they try, telling them we're ignoring it. Otherwise, they might be confused to see cavy-cli sat there.

**Review by AbigailMcP — APPROVED** [review]

**Pull request merged**